jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java

```diff
@@ -561,10 +635,19 @@ public String getPath()
         return _path;
     }
 
+    /**
+     * @return The decoded canonical path.
+     * @see URIUtil#canonicalPath(String)
+     */
     public String getDecodedPath()
     {
         if (_decodedPath == null && _path != null)
-            _decodedPath = URIUtil.decodePath(_path);
+        {
+            String canonical = URIUtil.canonicalPath(_path);
+            if (canonical == null)
+                throw new BadMessageException("Bad URI");
+            _decodedPath = URIUtil.decodePath(canonical);
+        }
```

Comment:
I don't like throwing when calling this method.
If the `HttpURI` is invalid, we should fail at `parse()` and not even create the `HttpURI` object, otherwise it seems a good object until I happen to call `getDecodedPath()` where it throws because it was bad.

Reply:
It is done here to be lazy in case the decoded path is never needed, but I guess it is fine to spend a little more effort on the minority of URIs that actually need to be decoded.

Reply:
@sbordet this has kind of resulted in a bigger refactor of this class than just fixing the problem... but I think it is more consistent now, so like it better.... we just need to really carefully review.
parse now always sets the decodedPath. The other big change is that most of the constructors now call parse, even if just to parse the path. Thus we are checking for ambiguous paths and bad canonical paths in more situations.

Reply:
@sbordet Note also that we have refactored this class in 10 to be immutable, so if we fix too much more here the merge is going to be difficult.